### PR TITLE
feat: add duplicate_of support

### PR DESCRIPTION
**Pull request by johnny243**
For #20

### CODE REVIEW

File: lib/protocol/payloads/functions.ts [hunk 1]
```diff
@@ -65,6 +65,7 @@ export async function PayloadsByDuplicating(
       conflict_of: payload.uuid,
     };
   }
+  override.duplicate_of = payload.uuid;
```
Comment (mobitar) [inline lib/protocol/payloads/functions.ts:68]: This should be defined under line 60.

File: lib/protocol/payloads/pure_payload.ts [hunk 2]
```diff
@@ -119,6 +120,8 @@ export class PurePayload {
       this.version = (this.content as PayloadContent).version;
     }
 
+    this.duplicate_of = rawPayload.duplicate_of;
```
Comment (mobitar) [inline lib/protocol/payloads/pure_payload.ts:123]: Should be under line 100.

**Comment by mobitar** [comment]
I forgot to mention that we also want to send this field to the server, save it to disk payloads, and include in backup files the user downloads:

- https://github.com/standardnotes/snjs/blob/5df6851ea4a59b17503ab99991d3d71b3c5d6cf8/lib/protocol/payloads/generator.ts#L121
- https://github.com/standardnotes/snjs/blob/5df6851ea4a59b17503ab99991d3d71b3c5d6cf8/lib/protocol/payloads/generator.ts#L93
- https://github.com/standardnotes/snjs/blob/5df6851ea4a59b17503ab99991d3d71b3c5d6cf8/lib/protocol/payloads/generator.ts#L104

Do we have tests that ensures certain fields are included in certain ejectable situations? For example, if syncing, items_key_id and duplicate_of (amongst others) should be in [ejected](https://github.com/standardnotes/snjs/blob/5df6851ea4a59b17503ab99991d3d71b3c5d6cf8/lib/protocol/payloads/pure_payload.ts#L133) payload.

**Comment by johnny243** [comment]
> I forgot to mention that we also want to send this field to the server, save it to disk payloads, and include in backup files the user downloads:

Done :+1: 

> Do we have tests that ensures certain fields are included in certain ejectable situations? For example, if syncing, items_key_id and duplicate_of (amongst others) should be in [ejected](https://github.com/standardnotes/snjs/blob/5df6851ea4a59b17503ab99991d3d71b3c5d6cf8/lib/protocol/payloads/pure_payload.ts#L133) payload.

I found [this](https://github.com/standardnotes/snjs/blob/4a8577bc148fbcbd0d775d9a55606b54f71dbec1/test/protocol.test.js#L79) test, which is just making sure there are no meta-data like `fields`, `source` and `format` when the payload is `ejected()`.

In order to write tests for this situations, should I use the `payloadsByEncryptingPayload` method from `protocolService` and pass the corresponding [EncryptionIntent](https://github.com/standardnotes/snjs/blob/004/lib/protocol/intents.ts) which represents the situation?

Example:

```javascript
const payload = { /* Generate payload */ };
const encryptedPayload = await this.application.protocolService.payloadByEncryptingPayload(
  payload,
  EncryptionIntent.Sync // For the syncing situation
);
const ejected = encryptedPayload.ejected();
expect(ejected).to.be.ok
expect(ejected).to.have.keys('items_key_id', 'duplicate_of', /* ... and other expected fields to be ejected */);
```

**Comment by mobitar** [comment]
In this case you would be testing the function `payloadByEncryptingPayload` to ensure it produces the correct result. Which perhaps should be a test of its own. But for now you should specifically target the underlying function that `payloadByEncryptingPayload` uses to create payloads.

**Pull request merged**